`bacalhau logs` panics (always?)

```
% go run . logs --simon 
panic: unable to redefine 'h' shorthand in "logs" flagset: it's already used for "history" flag

goroutine 1 [running]:
github.com/spf13/pflag.(*FlagSet).AddFlag(0x140019bff00, 0x140019cd860)
        /Users/simon/go/pkg/mod/github.com/spf13/pflag@v1.0.5/flag.go:874 +0x354
github.com/spf13/pflag.(*FlagSet).VarPF(0x14001315b98?, {0x107425a80, 0x140018bdc6e}, {0x106372ea5, 0x4}, {0x106bf1de0, 0x1}, {0x140018bdc61, 0xd})
        /Users/simon/go/pkg/mod/github.com/spf13/pflag@v1.0.5/flag.go:831 +0x148
github.com/spf13/pflag.(*FlagSet).BoolVarP(0x106f649a0?, 0x1063a5306?, {0x106372ea5?, 0x1063a573e?}, {0x106bf1de0?, 0x1?}, 0x1?, {0x140018bdc61?, 0x0?})
        /Users/simon/go/pkg/mod/github.com/spf13/pflag@v1.0.5/bool.go:55 +0x48
github.com/spf13/pflag.(*FlagSet).BoolP(...)
        /Users/simon/go/pkg/mod/github.com/spf13/pflag@v1.0.5/bool.go:80
github.com/spf13/cobra.(*Command).InitDefaultHelpFlag(0x1?)
        /Users/simon/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:1148 +0x10c
github.com/spf13/cobra.(*Command).execute(0x140019afb00, {0x1400196b490, 0x1, 0x1})
        /Users/simon/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:858 +0xb0
github.com/spf13/cobra.(*Command).ExecuteC(0x1400198fb00)
        /Users/simon/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:1068 +0x340
github.com/spf13/cobra.(*Command).Execute(...)
        /Users/simon/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:992
github.com/bacalhau-project/bacalhau/cmd/cli.Execute()
        /Users/simon/bacalhau/cmd/cli/root.go:196 +0x334
main.main()
        /Users/simon/bacalhau/main.go:37 +0x140
```

Something to do with it trying to automatically generate a help flag when the usage is printed.

--- a/cmd/cli/logs/logs.go
+++ b/cmd/cli/logs/logs.go
@@ -49,10 +49,5 @@ func NewCmd() *cobra.Command {
 		`Follow the logs in real-time after retrieving the current logs.`,
 	)
 
-	logsCmd.PersistentFlags().BoolVarP(
-		&options.Follow, "history", "h", false,
-		`Show all logs with history. TODO better definition`,
-	)
-
 	return logsCmd
 }

--- a/cmd/util/fatal.go
+++ b/cmd/util/fatal.go
@@ -14,7 +14,7 @@ func fatalError(cmd *cobra.Command, err error, code int) {
 		if !strings.HasSuffix(msg, "\n") {
 			msg += "\n"
 		}
-		cmd.Print(msg)
+		cmd.PrintErr(msg)
 	}
 	os.Exit(code)
 }

--- a/cmd/util/logging.go
+++ b/cmd/util/logging.go
@@ -2,6 +2,8 @@ package util
 
 import (
 	"context"
+	"encoding/json"
+	"errors"
 	"fmt"
 	"os"
 	"os/signal"
@@ -88,8 +90,9 @@ func ApplyPorcelainLogLevel(cmd *cobra.Command, _ []string) {
 }
 
 type Msg struct {
-	Tag  uint8
-	Data string
+	Tag          uint8
+	Data         string
+	ErrorMessage string
 }
 
 func readLogoutput(ctx context.Context, cmd *cobra.Command, conn *websocket.Conn) error {
@@ -112,10 +115,27 @@ func readLogoutput(ctx context.Context, cmd *cobra.Command, conn *websocket.Conn
 			if err != nil {
 				// If the error is NOT a CloseNormal then log the error
 				if websocket.IsUnexpectedCloseError(err, websocket.CloseNormalClosure) {
-					cmd.PrintErrf("failed to read: %s", err)
+					cmd.PrintErrf("Error: failed to read message: %s", err)
 				}
 
 				exiting = true
+				continue
+			}
+
+			if msg.ErrorMessage != "" {
+				var errResponse bacerrors.ErrorResponse
+				err := json.Unmarshal([]byte(msg.ErrorMessage), &errResponse)
+				if err != nil {
+					Fatal(cmd, fmt.Errorf("failed decoding error message from server: %s", err), 1)
+				}
+
+				e := fmt.Sprintf("Error: %s", &errResponse)
+				Fatal(cmd, errors.New(e), 1)
+
+				_ = conn.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(websocket.CloseNormalClosure, ""))
+				conn.Close()
+
+				return
 			}
 
 			if msg.Tag == 1 {

--- a/pkg/requester/publicapi/endpoints_logs.go
+++ b/pkg/requester/publicapi/endpoints_logs.go
@@ -22,8 +22,9 @@ import (
 type logRequest = publicapi.SignedRequest[model.LogsPayload] //nolint:unused // Swagger wants this
 
 type Msg struct {
-	Tag  uint8
-	Data string
+	Tag          uint8
+	Data         string
+	ErrorMessage string
 }
 
 // logs godoc
@@ -61,7 +62,7 @@ func (s *RequesterAPIServer) logs(res http.ResponseWriter, req *http.Request) {
 	err = conn.ReadJSON(&srequest)
 	if err != nil {
 		errorResponse := bacerrors.ErrorToErrorResponse(errors.Errorf("error reading signed request: %s", err))
-		_ = conn.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(websocket.CloseAbnormalClosure, errorResponse))
+		s.writeErrorMessage(ctx, conn, errorResponse)
 		return
 	}
 
@@ -72,7 +73,8 @@ func (s *RequesterAPIServer) logs(res http.ResponseWriter, req *http.Request) {
 	buffer := bytes.NewReader(srequest)
 	payload, err := publicapi.UnmarshalSigned[model.LogsPayload](ctx, buffer)
 	if err != nil {
-		_ = conn.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(websocket.CloseAbnormalClosure, "failed to decode request"))
+		errorResponse := bacerrors.ErrorToErrorResponse(errors.New("failed to decode request"))
+		s.writeErrorMessage(ctx, conn, errorResponse)
 		return
 	}
 
@@ -81,8 +83,8 @@ func (s *RequesterAPIServer) logs(res http.ResponseWriter, req *http.Request) {
 	// Get the job, check it exists and check it belongs to the same client
 	job, err := s.jobStore.GetJob(ctx, payload.JobID)
 	if err != nil {
-		log.Ctx(ctx).Debug().Msgf("Missing job: %s", err)
-		_ = conn.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(websocket.CloseAbnormalClosure, err.Error()))
+		errorResponse := bacerrors.ErrorToErrorResponse(errors.Errorf("failed to find job: %s", payload.JobID))
+		s.writeErrorMessage(ctx, conn, errorResponse)
 		return
 	}
 
@@ -93,8 +95,8 @@ func (s *RequesterAPIServer) logs(res http.ResponseWriter, req *http.Request) {
 		log.Ctx(ctx).Debug().Msgf("Mismatched ClientIDs for logs, existing job: %s and log request: %s",
 			job.Metadata.ClientID, payload.ClientID)
 
-		errorResponse := bacerrors.ErrorToErrorResponse(errors.Errorf("mismatched client id: %s", payload.ClientID))
-		_ = conn.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(websocket.CloseAbnormalClosure, errorResponse))
+		errorResponse := bacerrors.ErrorToErrorResponse(errors.Errorf("invalid client id: %s", payload.ClientID))
+		s.writeErrorMessage(ctx, conn, errorResponse)
 		return
 	}
 
@@ -107,7 +109,7 @@ func (s *RequesterAPIServer) logs(res http.ResponseWriter, req *http.Request) {
 	response, err := s.requester.ReadLogs(ctx, logRequest)
 	if err != nil {
 		errorResponse := bacerrors.ErrorToErrorResponse(errors.Errorf("read logs failure: %s", err))
-		_ = conn.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(websocket.CloseAbnormalClosure, errorResponse))
+		s.writeErrorMessage(ctx, conn, errorResponse)
 		return
 	}
 
@@ -119,7 +121,7 @@ func (s *RequesterAPIServer) logs(res http.ResponseWriter, req *http.Request) {
 	client, err := logstream.NewLogStreamClient(ctx, response.Address)
 	if err != nil {
 		errorResponse := bacerrors.ErrorToErrorResponse(errors.Errorf("logstream client create failure: %s", err))
-		_ = conn.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(websocket.CloseInternalServerErr, errorResponse))
+		s.writeErrorMessage(ctx, conn, errorResponse)
 		return
 	}
 	defer client.Close()
@@ -127,7 +129,7 @@ func (s *RequesterAPIServer) logs(res http.ResponseWriter, req *http.Request) {
 	err = client.Connect(ctx, payload.ExecutionID, payload.WithHistory, payload.Follow)
 	if err != nil {
 		errorResponse := bacerrors.ErrorToErrorResponse(errors.Errorf("logstream connect failure: %s", err))
-		_ = conn.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(websocket.CloseInternalServerErr, errorResponse))
+		s.writeErrorMessage(ctx, conn, errorResponse)
 		return
 	}
 
@@ -138,8 +140,7 @@ func (s *RequesterAPIServer) logs(res http.ResponseWriter, req *http.Request) {
 			break
 		}
 		if err != nil {
-			log.Ctx(ctx).Error().Msgf("Stream read failure: %s", err)
-			_ = conn.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(websocket.CloseNormalClosure, ""))
+			log.Ctx(ctx).Error().Err(err).Msgf("Stream read failure. May be reset?: %s", err)
 			break
 		}
 
@@ -151,6 +152,17 @@ func (s *RequesterAPIServer) logs(res http.ResponseWriter, req *http.Request) {
 	}
 
 	_ = conn.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(websocket.CloseNormalClosure, ""))
+}
+
+func (s *RequesterAPIServer) writeErrorMessage(ctx context.Context, conn *websocket.Conn, errorMsg string) {
+	msg := Msg{
+		ErrorMessage: errorMsg,
+	}
+
+	err := conn.WriteJSON(msg)
+	if err != nil {
+		log.Ctx(ctx).Error().Msgf("websocket write failure sending error: %s", err)
+	}
 }
 
 func (s *RequesterAPIServer) writeDataFrame(ctx context.Context, conn *websocket.Conn, frame logger.DataFrame) error {

